core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -2915,6 +2966,84 @@ wasi_ssp_sock_bind(
     return __WASI_ESUCCESS;
 }
 
+__wasi_errno_t
+wasi_ssp_addr_resolve(
+#if !defined(WASMTIME_SSP_STATIC_CURFDS)
+    struct fd_table *curfds,
+#endif
+    const char *host, const char *service, __wasi_addr_info_hints_t *hints,
+    __wasi_addr_info_t *addr_info, __wasi_size_t addr_info_size,
+    __wasi_size_t *max_size)
+{
+    struct addrinfo posix_hints, *res, *result;
+    __wasi_size_t pos = 0;
+    int ret;
+
+    if (hints->hints_enabled) {
+        memset(&posix_hints, 0, sizeof(posix_hints));
+        posix_hints.ai_family = hints->family == INET4 ? PF_INET : PF_INET6;
+        posix_hints.ai_socktype =
+            hints->type == SOCKET_STREAM ? SOCK_STREAM : SOCK_DGRAM;
+    }
+
+    ret = getaddrinfo(host, strlen(service) == 0 ? NULL : service,
+                      hints->hints_enabled ? &posix_hints : NULL, &result);
+    if (ret != 0) {
+        return convert_getaddrinfo_error_to_wasi_error_code(ret);
+    }
+
+    res = result;
+    while (res) {
+        if (addr_info_size > pos) {
+            if (
+                // Allow only IPv4 and IPv6
+                (res->ai_family != AF_INET && res->ai_family != AF_INET6)
+                // Allow only UDP and TCP
+                || (res->ai_socktype != SOCK_DGRAM
+                    && res->ai_socktype != SOCK_STREAM)
+                || (res->ai_protocol != IPPROTO_TCP
+                    && res->ai_protocol != IPPROTO_UDP)) {
+                res = res->ai_next;
+                continue;
+            }
+
+            switch (res->ai_family) {
```

Comment:
since we have already reduced values of `res->ai_family` to  either `AF_INET` or `AF_INET6`, maybe a smile `if...else..` is enough